File: packages/roster-client/src/Harvest.cs
using System.Collections.Generic;

namespace MadDr.RosterClient
{
    /// <summary>
    /// Harvest & carry -- the display-side C# twin of genome-core's
    /// harvest.ts (docs/22 harvester morphology), the same discipline as
    /// Locomotion/Weapon: the Lab and the Unity battlefield must agree on
    /// how much a given genome gathers, carries, and how much its load
    /// slows it, or the same harvester would behave differently in the two
    /// places it appears. Verified against golden values captured from the
    /// real harvest.ts running in node (Tests~/HarvestTests.cs).
    ///
    /// Nothing here is a new gene: gather comes from the HAND family, carry
    /// capacity from a SENSOR-slot storage vessel plus body bulk, and
    /// weight slows the carrier -- twice as hard for flyers (winged /
    /// floater), because flight pays for every unit carried.
    /// </summary>
    public sealed class HarvestProfile
    {
        public double GatherBlood { get; }
        public double GatherBone { get; }
        public double GatherBrain { get; }
        /// <summary>The hand tool can drain a target that is still ALIVE
        /// (the lamprey/siphon fantasy), not just corpses.</summary>
        public bool DrainsLiving { get; }
        /// <summary>Total onboard carry capacity, all lanes pooled.</summary>
        public double Capacity { get; }
        /// <summary>The sensor slot carries a dedicated storage vessel.</summary>
        public bool HasVessel { get; }
        /// <summary>winged/floater plans -- the ones the weight rule bites.</summary>
        public bool Flies { get; }

        public HarvestProfile(double gatherBlood, double gatherBone, double gatherBrain,
            bool drainsLiving, double capacity, bool hasVessel, bool flies)
        {
            GatherBlood = gatherBlood;
            GatherBone = gatherBone;
            GatherBrain = gatherBrain;
            DrainsLiving = drainsLiving;
            Capacity = capacity;
            HasVessel = hasVessel;
            Flies = flies;
        }
    }

    public static class Harvest
    {
        // Gather-rate multipliers by hand family, per resource lane -- must
        // match harvest.ts HARVEST_TOOLS. Families absent here use TOOLLESS.
        private struct Tool
        {
            public double Blood, Bone, Brain;
            public bool Living;
            public Tool(double bl, double bo, double br, bool living) { Blood = bl; Bone = bo; Brain = br; Living = living; }
        }

        private static readonly Tool Toolless = new Tool(0.4, 0.4, 0.2, false);

        private static readonly Dictionary<string, Tool> Tools = new Dictionary<string, Tool>
        {
            { "lamprey_maw", new Tool(3.0, 0.3, 0.4, true) },
            { "ichor_siphon", new Tool(2.4, 0.3, 0.8, true) },
            { "bone_saw", new Tool(0.5, 3.0, 0.6, false) },
            { "claw_hand", new Tool(1.0, 1.0, 0.5, false) },
            { "pincer", new Tool(0.8, 1.4, 0.5, false) },
            { "chain_blade", new Tool(0.7, 1.8, 0.3, false) },
            { "tentacle", new Tool(1.2, 0.4, 0.6, false) },
        };

        // Storage-vessel families (sensor homolog) and full-expression
        // capacity -- must match harvest.ts STORAGE_FAMILIES.
        private static readonly Dictionary<string, double> Storage = new Dictionary<string, double>
        {
            { "storage_bladder", 60 },
            { "steel_tank", 70 },
            { "amber_vesicle", 55 },
        };

        // v0.1 tuning knobs (mirror harvest.ts) ------------------------------
        private const double BaseCapacity = 10;
        private const double BulkCapacity = 15;
        private const double BlobBonus = 1.5;
        private const double GroundPenalty = 0.25;
        private const double FlightPenalty = 0.5;
        private const double GroundFloor = 0.6;
        private const double FlightFloor = 0.4;

        public static HarvestProfile Profile(GenomeDto g)
        {
            var plan = g.Body.Plan;
            var ignored = IgnoredSlots(plan);

            // gather: from the hand tool, scaled by its expressed working size
            double gb = 0, go = 0, gr = 0;
            var drainsLiving = false;
            if (!ignored.Contains("hand"))
            {
                var hand = g.Slots.Hand;
                if (!IsVestigial(hand.Family))
                {
                    Tool spec;
                    if (!Tools.TryGetValue(hand.Family, out spec)) spec = Toolless;
                    var length = Express(hand.Family, Axis.Length, At(hand.Params, 0, 0.5));
                    var girth = Express(hand.Family, Axis.Girth, At(hand.Params, 1, 0.5));
                    var size = 0.5 + (length + girth) / 2.0; // [0.5, 1.5]
                    gb = spec.Blood * size;
                    go = spec.Bone * size;
                    gr = spec.Brain * size;
                    drainsLiving = spec.Living;
                }
            }

            // capacity: body bulk + the storage vessel, times the blob bonus
            var capacity = BaseCapacity + BulkCapacity * At(g.Body.Params, 1, 0.5);
            var hasVessel = false;
            if (!ignored.Contains("sensor"))
            {
                var sensor = g.Slots.Sensor;
                double vesselMax;
                if (Storage.TryGetValue(sensor.Family, out vesselMax))
                {
                    hasVessel = true;
                    var length = Express(sensor.Family, Axis.Length, At(sensor.Params, 0, 0.5));
                    var girth = Express(sensor.Family, Axis.Girth, At(sensor.Params, 1, 0.5));
                    capacity += vesselMax * ((length + girth) / 2.0);
                }
            }
            if (plan == "blob") capacity *= BlobBonus;

            return new HarvestProfile(gb, go, gr, drainsLiving, capacity, hasVessel, Flies(plan));
        }

        /// <summary>Speed multiplier for a carrier at `fill` (0..1) on the
        /// ground -- linear, floored: laden trudging, never stalling.</summary>
        public static double GroundSpeedFactor(double fill)
        {
            var f = Clamp01(fill);
            var factor = 1 - GroundPenalty * f;
            return factor < GroundFloor ? GroundFloor : factor;
        }

        /// <summary>Speed multiplier for a FLYING carrier at `fill` -- twice
        /// the ground sensitivity (every unit aloft is paid for) but
        /// floored: a full tank makes a slow flyer, never a grounded one
        /// (docs/22 SS1 never-annoying contract).</summary>
        public static double FlightSpeedFactor(double fill)
        {
            var f = Clamp01(fill);
            var factor = 1 - FlightPenalty * f;
            return factor < FlightFloor ? FlightFloor : factor;
        }

        public static bool Flies(string plan)
        {
            return plan == "winged" || plan == "floater";
        }

        // ---- canalized expression (subset of catalog.ts bounds) -------------
        // Only length/girth are read here, and only families whose bounds
        // DIFFER from the [0,1] default need an entry; everything else
        // expresses as identity. Kept in lockstep with catalog.ts by the
        // golden test.
        private enum Axis { Length, Girth }

        private static double Express(string family, Axis axis, double gene)
        {
            double lo = 0, hi = 1;
            switch (family)
            {
                case "lamprey_maw": if (axis == Axis.Girth) { lo = 0.25; hi = 0.85; } break;
                case "ichor_siphon": if (axis == Axis.Girth) { lo = 0.2; hi = 0.7; } break;
                case "bone_saw": if (axis == Axis.Girth) { lo = 0.3; hi = 0.9; } break;
                case "chain_blade": if (axis == Axis.Girth) { lo = 0.3; hi = 0.9; } break;
                case "tentacle": if (axis == Axis.Girth) { lo = 0.0; hi = 0.7; } break;
                case "spore_launcher": if (axis == Axis.Girth) { lo = 0.35; hi = 1.0; } break;
                case "photon_blaster": if (axis == Axis.Girth) { lo = 0.4; hi = 1.0; } break;
                case "storage_bladder":
                    if (axis == Axis.Girth) { lo = 0.4; hi = 1.0; } else { lo = 0.2; hi = 0.8; }
                    break;
                case "steel_tank":
                    if (axis == Axis.Girth) { lo = 0.35; hi = 0.9; } else { lo = 0.3; hi = 0.9; }
                    break;
                case "amber_vesicle": if (axis == Axis.Girth) { lo = 0.3; hi = 0.9; } break;
            }
            return lo + gene * (hi - lo);
        }

        private static bool IsVestigial(string family)
        {
            return family == "hand_stump" || family == "sensor_stub"
                || family == "eye_socket" || family == "leg_stump";
        }

        private static HashSet<string> IgnoredSlots(string plan)
        {
            switch (plan)
            {
                case "blob": return new HashSet<string> { "leg" };
                case "serpentine": return new HashSet<string> { "leg" };
                case "treant": return new HashSet<string> { "leg" };
                case "floater": return new HashSet<string> { "leg" };
                default: return new HashSet<string>();
            }
        }

        private static double At(double[] arr, int i, double fallback)
        {
            return arr != null && i < arr.Length ? arr[i] : fallback;
        }

        private static double Clamp01(double x)
        {
            return x < 0 ? 0 : x > 1 ? 1 : x;
        }
    }
}

File: packages/genome-core/src/harvest.ts
/**
 * Harvest & carry -- the derived stats behind harvester creatures
 * (docs/22 harvester morphology). Pure functions of the genome, exactly
 * like energy.ts: nothing here is a new gene, everything is EXPRESSED
 * from parts the creature already carries, so harvesters are bred and
 * grafted in the Lab like any other design -- never a bolt-on unit type.
 *
 *   gather   -- how fast the creature strips blood/bone/brains from a
 *               target, set by its HAND family (a lamprey maw drinks, a
 *               bone saw cuts, ordinary claws paw uselessly slowly)
 *   capacity -- how much it can carry, set by its SENSOR-slot storage
 *               vessel (bladder / steel tank / amber vesicles) plus the
 *               body's own bulk; the blob plan gets a plan bonus -- an
 *               amorphous body IS a bag
 *   weight   -- carried load slows the carrier: mildly on the ground,
 *               twice as hard in the air (winged/floater plans), because
 *               flight pays for every kilogram (creator direction)
 *
 * All numbers v0.1 -- Phase-2 sandbox tuning (docs/22 SS10). Penalty
 * FLOORS follow docs/22 SS1's never-annoying contract: a fully-laden
 * flyer is slow, never grounded; a laden walker trudges, never stalls.
 */

import {
  SLOT_NAMES,
  bodyAxis,
  partAxis,
  type Genome,
} from "./genome.js";
import { BODY_PLANS, express, isVestigial } from "./catalog.js";

/** Gather-rate multipliers by hand family, per resource lane. Families
 * absent from this table use TOOLLESS -- anything with a working hand
 * can tear at a corpse, badly. `drainsLiving`: the tool can pull blood
 * from a target that is still alive (the lamprey/siphon fantasy);
 * everything else harvests corpses only. */
export interface GatherSpec {
  readonly blood: number;
  readonly bone: number;
  readonly brain: number;
  readonly drainsLiving: boolean;
}

const TOOLLESS: GatherSpec = { blood: 0.4, bone: 0.4, brain: 0.2, drainsLiving: false };

export const HARVEST_TOOLS: Readonly<Record<string, GatherSpec>> = {
  // dedicated harvest tools (docs/22)
  lamprey_maw: { blood: 3.0, bone: 0.3, brain: 0.4, drainsLiving: true },
  ichor_siphon: { blood: 2.4, bone: 0.3, brain: 0.8, drainsLiving: true },
  bone_saw: { blood: 0.5, bone: 3.0, brain: 0.6, drainsLiving: false },
  // general-purpose hands that happen to cut and grab decently
  claw_hand: { blood: 1.0, bone: 1.0, brain: 0.5, drainsLiving: false },
  pincer: { blood: 0.8, bone: 1.4, brain: 0.5, drainsLiving: false },
  chain_blade: { blood: 0.7, bone: 1.8, brain: 0.3, drainsLiving: false },
  tentacle: { blood: 1.2, bone: 0.4, brain: 0.6, drainsLiving: false },
};

/** Storage-vessel families (sensor homolog) and the capacity each grants
 * at full expression. Not a stat on other sensors: an antenna senses, a
 * bladder carries -- the slot is the trade. */
export const STORAGE_FAMILIES: Readonly<Record<string, number>> = {
  storage_bladder: 60,
  steel_tank: 70,
  tank_backpack: 85, // two tanks beat one, short of a flat doubling
  amber_vesicle: 55,
};

// v0.1 tuning knobs ----------------------------------------------------------
const BASE_CAPACITY = 10; // any creature can gulp a little
const BULK_CAPACITY = 15; // a bigger body holds more, vessel or not
const BLOB_BONUS = 1.5; // the amorphous plan is its own storage bag
const GROUND_PENALTY = 0.25; // fully laden ground speed loss
const FLIGHT_PENALTY = 0.5; // fully laden flight speed loss -- weight counts double aloft
const GROUND_FLOOR = 0.6; // never-annoying floors (docs/22 SS1)
const FLIGHT_FLOOR = 0.4;

/** Plans that fly (the weight rule's audience). A plan property, like
 * amphibious -- not a gene. */
const FLYING_PLANS: readonly string[] = ["winged", "floater"];

export interface HarvestProfile {
  /** Gather rates in resource-units/second against a valid target,
   * already scaled by the tool's expressed size. */
  readonly gather: { readonly blood: number; readonly bone: number; readonly brain: number };
  /** Whether the hand tool can drain a LIVING target. */
  readonly drainsLiving: boolean;
  /** Total onboard carry capacity (resource units, all lanes pooled). */
  readonly capacity: number;
  /** True when the sensor slot carries a dedicated storage vessel. */
  readonly hasVessel: boolean;
  /** True for winged/floater plans -- the ones the weight rule bites. */
  readonly flies: boolean;
}

export function harvestProfile(g: Genome): HarvestProfile {
  const ignored = BODY_PLANS[g.body.plan]?.ignoresSlots ?? [];

  // gather: from the hand tool, scaled by its expressed working size
  let gather = { blood: 0, bone: 0, brain: 0 };
  let drainsLiving = false;
  if (!ignored.includes("hand" as (typeof SLOT_NAMES)[number])) {
    const hand = g.slots.hand;
    if (!isVestigial(hand.family)) {
      const spec = HARVEST_TOOLS[hand.family] ?? TOOLLESS;
      const length = express(hand.family, "length", partAxis(hand, "length"));
      const girth = express(hand.family, "girth", partAxis(hand, "girth"));
      const size = 0.5 + (length + girth) / 2; // [0.5, 1.5]
      gather = {
        blood: spec.blood * size,
        bone: spec.bone * size,
        brain: spec.brain * size,
      };
      drainsLiving = spec.drainsLiving;
    }
  }

  // capacity: body bulk + the storage vessel, times the blob plan bonus
  let capacity = BASE_CAPACITY + BULK_CAPACITY * bodyAxis(g.body, "bulk");
  let hasVessel = false;
  if (!ignored.includes("sensor" as (typeof SLOT_NAMES)[number])) {
    const sensor = g.slots.sensor;
    const vesselMax = STORAGE_FAMILIES[sensor.family];
    if (vesselMax !== undefined) {
      hasVessel = true;
      const length = express(sensor.family, "length", partAxis(sensor, "length"));
      const girth = express(sensor.family, "girth", partAxis(sensor, "girth"));
      capacity += vesselMax * ((length + girth) / 2);
    }
  }
  if (g.body.plan === "blob") capacity *= BLOB_BONUS;

  return {
    gather,
    drainsLiving,
    capacity,
    hasVessel,
    flies: FLYING_PLANS.includes(g.body.plan),
  };
}

/** Speed multiplier for a carrier at `fill` (0 = empty, 1 = full),
 * on the ground. Linear, floored -- laden trudging, never stalling. */
export function groundSpeedFactor(fill: number): number {
  const f = fill < 0 ? 0 : fill > 1 ? 1 : fill;
  const factor = 1 - GROUND_PENALTY * f;
  return factor < GROUND_FLOOR ? GROUND_FLOOR : factor;
}

/** Speed multiplier for a FLYING carrier at `fill`. Twice the ground
 * sensitivity -- every kilogram aloft is paid for -- but floored: a full
 * tank makes a slow flyer, never a grounded one (docs/22 SS1). */
export function flightSpeedFactor(fill: number): number {
  const f = fill < 0 ? 0 : fill > 1 ? 1 : fill;
  const factor = 1 - FLIGHT_PENALTY * f;
  return factor < FLIGHT_FLOOR ? FLIGHT_FLOOR : factor;
}
